Add: Independent scrolling in documentation sidebar

vihaanthora opened the pull request:
This pull request closes task https://issues.jenkins.io/browse/WEBSITE-799.

It introduces independent scrolling in the documentation sidebar, which is also made collapsible to facilitate the mobile experience. 

Upon expanding the sidebar, the main content gets pushed rightwards, and the footer is shrunk adequately. 

### An Alternative Approach
Upon expanding, we can overlap the sidebar on the content. This will make the toggling much faster for pages with large content.

Tagging the mentors - @kwhetstone @arpoch

zbynek commented:
This affects many pages other then the step reference (e.g. https://deploy-preview-5227--jenkins-io-site-pr.netlify.app/doc/book/pipeline/syntax/ ) and it's collapsed by default, making it hard to find the step reference in the first place.

vihaanthora commented:
This particular change was aimed at the entire documentation (not just the Steps Reference page), as there are several other pages that have a lot of content and can benefit from a static sidebar. As far as the collapsed sidebar is concerned, we can change the default behavior to be expanded when the page is loaded, but that might not be suitable for devices with smaller screen widths.

kwhetstone commented:
I've heard some good feedback about this feature.  It looks like the most important thing here is to start with the sidebar open by default, and keep the state between pages.

I'm looking at the preview site on this page: [security book](https://deploy-preview-5227--jenkins-io-site-pr.netlify.app/doc/book/security/) and it looks like there's a lot of white space in the sidebar.  Is that just a function of the preview environment or is it actually something that's happening with the page layout?

kwhetstone reviewed (commented):
Should have left my comment as a "comment" so see above

Also, is there an icon in the Jenkins website we could use for the arrow?  (someone mentioned matching icons)  I'm just not sure at this time, and that would be more about look and feel.

vihaanthora commented:
Could you please share a screenshot and mark the whitespace that you are referring to? Because as far as the width of the sidebar is concerned, it should be the same across all the pages.

kwhetstone commented:
Reflecting on this it might be an alignment thing.  It looks like the content of the sidebar is aligned with the left side vs how close it is to the content currently on the website.

I think is is probably what's happening with the footer not extending all the way across the bottom of the page here: 
![image](https://user-images.githubusercontent.com/14875282/174000227-39ad67d2-193d-4873-a118-255ea74c58d4.png)

vihaanthora commented:
I think this is to do with the way I have set the ratios. It looks fine on my machine, but it is getting stretched out on a monitor. 

<img width="960" alt="image" src="https://user-images.githubusercontent.com/77713537/174001024-a300923a-bd61-4b8a-ad65-606f57cbf6d7.png">

vihaanthora commented:
If the footer shrinkage is not desirable, can we overlap the sidebar over it whenever it is expanded?

timja commented:
I have no sidebar on https://deploy-preview-5227--jenkins-io-site-pr.netlify.app/doc/developer/security/ 

this is not workable I cannot navigate

vihaanthora commented:
@timja the error is occuring because of class name conflicts, I will assign a new name to this sidebar, and the issue will be solved. As far as the design itself is concerned, I tried the fixed position wrt parent container method, but was not successful in that. I am working on making the look more consistent.

vihaanthora commented:
The reason why the kubernetes style sidebar might not be suitable for jenkins is that in some screen ratios (I tried with 820X1180), scrolling the sidebar to the bottom-most point requires the parent container (content) to be scrolled all the way down. So, for pages with a very large content, the user might need to scroll it all the way down just to see all the options of the sidebar.

timja commented:
>  that in some screen ratios

Please provide screenshots or a video showing the issue you are describing

Kubernetes docs seem to work very well on mobile to me

vihaanthora commented:
For iPad Air screen size:

<img width="247" alt="image" src="https://user-images.githubusercontent.com/77713537/174009622-0453010c-42fb-42d4-9590-a7e5d5640fad.png">

I have scrolled the sidebar down as much as possible, still, the entire sidebar is not visible.

timja commented:
Ok there's a very minor cut-off on that one size:
![image](https://user-images.githubusercontent.com/21194782/174010281-d7947993-0ffc-49df-9547-8715a0dd0970.png)

I wouldn't use that as a reason to discard the approach though

vihaanthora commented:
Right, then I think we can use a similar scrolling behavior (we can anyways get rid of the clipping by setting a max height of the sidebar).

vihaanthora commented:
I have added the scrolling type shown by the Kubernetes sidebar. I have also removed the option to collapse/expand. Please let me know if this sidebar is better than the previous one.

timja commented:
Desktop looks better now.

Mobile not the best:
![image](https://user-images.githubusercontent.com/21194782/174049686-b9488904-3ad9-441f-b26f-1b5b70e93a65.png)

That table of content should move and the side bar should be either at the end of the content or expanded via a menu like with Kubernetes.

Having it at the beginning means that a user has to scroll through a lot.

Having it at the beginning is a pre-existing issue though so not a blocker.

But it's now compressed and wasn't previously which should be fixed

![image](https://user-images.githubusercontent.com/21194782/174050007-9677e3d1-ecc6-4d1e-b4c7-83cb9b6ae15e.png)

vihaanthora commented:
I am not sure why the mobile view is getting zoomed out by default. I checked the master branch and even it is showing the mobile view in a zoomed-out manner by default. 

I opened the preview on my mobile phone and the zoom level by default was correct, so I think that it is not an issue. 

If it solves it, I can reduce the max-height of the sidebar for mobile view, so that a user can scroll directly to the content (only the max-height amount needs to be scrolled) without having to scroll through the entire sidebar.

arpoch commented:
The React.Js docshttps://reactjs.org/docs/getting-started.html, is a good example if the table of content needs to be handy in mobile view.

vihaanthora commented:
@arpoch thanks for pointing this out. It seems like a great way to perform the required task. I am thinking of doing that in a separate pull request to keep things straightforward and bite-sized for this one.

timja reviewed (approved):
seems fine, room for improvement in the future but better than before I think

MarkEWaite commented:
@vihaanthora this was merged about 8 hours ago, yet we're not seeing the unmoving table of contents on the left.  https://www.jenkins.io/doc/book/security/ behaving as you expect?

vihaanthora commented:
@MarkEWaite it seems that none of the changes are reflected on the website (tried clearing cache also). Just to be sure, I pulled from master and ran it locally, where it was working just fine.

vihaanthora commented:
The changes are now visible.

MarkEWaite commented:
They are visible to me as well. Thanks for the improvement @vihaanthora !
